soc/arm/nordic_nrf/validate_base_addresses.c

```diff
@@ -15,270 +15,101 @@
+ *
+ * - 'compat' is the part that comes after "nordic,nrf-" in a devicetree
+ *   compatible, with special characters converted to underscores
+ * - 'peripheral' specifies the address by NRF_<peripheral>_BASE
+ */
+#define CHECK_COMPAT(compat, mdk)					\
+	BUILD_ASSERT(							\
+		UTIL_OR(UTIL_NOT(DT_HAS_COMPAT(compat)),		\
+			((DT_NUM_INST(compat) == 1) &&			\
+			 (DT_REG_ADDR(DT_INST(0, compat)) ==		\
+			  (u32_t)(NRF_##mdk)))))
+/*
+ * Use this instead when there are multiple possible instances of an
+ * IP block.
+ *
+ * We can't use DT_INST APIs here because they do not map to our SoC
+ * peripheral instance numbers.
+ *
+ * - 'nl' is the devicetree node label you want to check, with
+ *    special characters converted to underscores
+ * - 'instance' is used to check the address as NRF_<instance>_BASE.
+ *   Note that some remapping is done above for things like SPIM0 -> SPI0,
+ *   TWIM0 -> TWI0, etc.
+ */
+#define CHECK_NODELABEL(lbl, mdk)					\
+	BUILD_ASSERT(							\
+		UTIL_OR(UTIL_NOT(DT_HAS_NODE(DT_NODELABEL(lbl))),	\
+			(DT_REG_ADDR(DT_NODELABEL(lbl)) ==		\
+			 (u32_t)(NRF_##mdk))))
+
+CHECK_COMPAT(nordic_nrf_adc, ADC);
+CHECK_COMPAT(nordic_nrf_cc310, CRYPTOCELL);
+CHECK_COMPAT(nordic_nrf_clock, CLOCK);
+CHECK_COMPAT(nordic_nrf_dppic, DPPIC);
+CHECK_COMPAT(nordic_nrf_ficr, FICR);
+CHECK_COMPAT(nordic_nrf_ipc, IPC);
+CHECK_COMPAT(nordic_nrf_kmu, KMU);
+CHECK_COMPAT(nordic_nrf_nrf51_flash_controller, NVMC);
+CHECK_COMPAT(nordic_nrf_nrf52_flash_controller, NVMC);
+CHECK_COMPAT(nordic_nrf_nrf53_flash_controller, NVMC);
+CHECK_COMPAT(nordic_nrf_nrf91_flash_controller, NVMC);
+CHECK_COMPAT(nordic_nrf_power, POWER);
+CHECK_COMPAT(nordic_nrf_qdec, QDEC);
+CHECK_COMPAT(nordic_nrf_regulators, REGULATORS);
+CHECK_COMPAT(nordic_nrf_rng, RNG);
+CHECK_COMPAT(nordic_nrf_saadc, SAADC);
+CHECK_COMPAT(nordic_nrf_spu, SPU);
+CHECK_COMPAT(nordic_nrf_temp, TEMP);
+CHECK_COMPAT(nordic_nrf_uicr, UICR);
+CHECK_COMPAT(nordic_nrf_usbd, USBD);
+CHECK_COMPAT(nordic_nrf_vmc, VMC);
+
+CHECK_NODELABEL(egu0, EGU0);
+CHECK_NODELABEL(egu1, EGU1);
+CHECK_NODELABEL(egu2, EGU2);
+CHECK_NODELABEL(egu3, EGU3);
+CHECK_NODELABEL(egu4, EGU4);
+CHECK_NODELABEL(egu5, EGU5);
+CHECK_NODELABEL(gpio0, P0);
+CHECK_NODELABEL(gpio1, P1);
+CHECK_NODELABEL(gpiote, GPIOTE);
+CHECK_NODELABEL(i2s0,  I2S0);
+CHECK_NODELABEL(pdm0, PDM0);
+CHECK_NODELABEL(pwm0, PWM0);
+CHECK_NODELABEL(pwm1, PWM1);
+CHECK_NODELABEL(pwm2, PWM2);
+CHECK_NODELABEL(pwm3, PWM3);
+CHECK_NODELABEL(rtc0, RTC0);
+CHECK_NODELABEL(rtc1, RTC1);
+CHECK_NODELABEL(rtc2, RTC2);
+CHECK_NODELABEL(spi0, SPI0);
+CHECK_NODELABEL(spi1, SPI1);
+CHECK_NODELABEL(spi2, SPI2);
+CHECK_NODELABEL(spi3, SPI3);
+CHECK_NODELABEL(spi4, SPI4);
+CHECK_NODELABEL(timer0, TIMER0);
+CHECK_NODELABEL(timer1, TIMER1);
+CHECK_NODELABEL(timer2, TIMER2);
+CHECK_NODELABEL(timer3, TIMER3);
+CHECK_NODELABEL(timer4, TIMER4);
+CHECK_NODELABEL(twi0, TWI0);
+CHECK_NODELABEL(twi1, TWI1);
+CHECK_NODELABEL(twi2, TWI2);
+CHECK_NODELABEL(twi3, TWI3);
```

Comment:
Now I'm worried that sanitycheck didn't catch these. I imagine it's because the nodes aren't enabled by default on any of those boards in a hello world configuration. I will try to enable all the instances with an overlay and retest.